keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/AutoDiffBroadcast.java

```diff
@@ -0,0 +1,60 @@
+package io.improbable.keanu.vertices.dbl.nonprobabilistic.diff;
+
+import io.improbable.keanu.tensor.TensorShape;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import lombok.experimental.UtilityClass;
+
+import java.util.Arrays;
+
+/**
+ * This class is meant to help with auto diff in operations that support implicit broadcasting. E.g. In
+ * addition/subtraction/multiplication/division scalar operands can be operated with non-scalar operands.
```

Comment:
Obviously it's not just scalars that can be broadcast and lead to implicit partial changes - it's also compatible Tensors - presumably given the naming that's still a bug we have to fix at some point?

Reply:
Those non-scalar broadcast aren't supported yet and strictly prohibited through the shape check in the operations that could broadcast. When we add that support then this code will need to be tweaked to support the new cases.

Reply:
Cool stuff - If you see that as a bigger refactor, no bother.  I wonder if there's anything we can do to mark this as something that will need fixing if we add broadcasting properly so this isn't missed?